### DISCUSSION TITLE
Include libffi for fast, efficient integration with dynamic c libraries

### DISCUSSION THREAD

**Discussion by phronmophobic** [babashka/babashka#1011] [Ideas]
**Is your feature request related to a problem? Please describe.**
The main options for adding an additional library to bb are:
1. Compile the library into bb itself
2. Use babashka pods

Including libffi would be an alternative to babashka pods. The main benefits of using libffi over pods are:
1. gain access to large, existing ecosystem of c libraries
2. library calls have less overhead.
3. clojure/java libraries can be compiled to shared libraries using native image which would allow more jvm libraries to be made available to bb a la carte.

**Describe the solution you'd like**
The basic idea is that you can load and call functions from a dynamic library at runtime using libffi.

**Describe alternatives you've considered**
Babashka pods already provide an alternative. However, there are many existing libraries that already are c ABI compatible. Additionally, using pods can have meaningful overhead for certain applications. 

**Additional context**
This is a pretty experimental idea, but I think it could unlock a lot of value. If this feature sounds interesting, I could start gathering more info to determine if it's actually a good fit. I would also be interested in doing the bulk of the integration work. I've been thinking about how to script UIs with membrane, and this seems like one of the better options.

I know that binary size is an important factor for bb. libffi.a on mac osx is 53kb and I assume there's some additional size required for integrating it.

It's also possible that `native-image` will include libffi-like functionality directly in the future.

**Comment by borkdude** [comment]
Yes, this sounds like a very useful addition! One edge case may be the static (musl) compiled version of bb, not sure if that can use libffi as well. Happy to collaborate with you on that.

**Comment by phronmophobic** [comment]
Awesome. I'll be on vacation for the next two weeks, but I'll hopefully have some time in the second half of this month.

**Comment by phronmophobic** [comment]
Initial proof of concept for an ffi interface, https://github.com/phronmophobic/babashka/tree/ffi.

The wrapper for libffi is here, https://github.com/phronmophobic/clj-libffi. Currently, it will only work on Mac OSX, but it shouldn't be too much work to add support for linux and windows.

Compiling with out including libffi was ~80mb and adding the libffi namespace increased it to about ~148mb. That seems like a lot, but it does provide access to a wide ecosystem of c libraries or java libraries compiled to shared libraries with native-image. Not sure where most of the bulk is coming from, but it might be possible to strip some of the dtype-next dependencies down or implement the interface directly to reduce the size.

```clojure
bash-3.2$ ./bb
Babashka v0.6.2-SNAPSHOT REPL.
Use :repl/quit or :repl/exit to quit the REPL.
Clojure rocks, Bash reaches.

user=> (require '[babashka.ffi :as ffi])
nil
user=> (ffi/call "cos" :float64 :float64 42)
-0.3999853149883513
user=> (ffi/load-library "/usr/lib/libobjc.A.dylib")
#object[tech.v3.datatype.ffi.Pointer 0x35d0b188 "{:address 0x0000000209E72990 }"]
user=> (ffi/call "objc_getClass" :pointer :pointer (ffi/string->c "NSArray"))
#object[tech.v3.datatype.ffi.Pointer 0x4cf726a "{:address 0x00007FFF883B8A70 }"]
```

Even if calling c libraries would be technically possible with this version, there's still open questions as how to make it easy for babashka users to obtain and call c libraries.

Always more to do! Any thoughts on what you see so far?

**Reply by borkdude** [reply]
@cnuernber This looks pretty interesting ^

**Reply by borkdude** [reply]
This is very awesome work. I will study it more in the weekend. The binary size is a bit concerning, but we can at the very least make an experimental SCI-based tool where we incorporate this and see how it evolves. Also with respect to building other Clojure libraries as shared libs. Would it be possible to pass function references over the native FFI border for example? I would be surprised if this was possible.

**Reply by phronmophobic** [reply]
> The binary size is a bit concerning, but we can at the very least make an experimental SCI-based tool where we incorporate this and see how it evolves.

Sounds good. It looks like many of the included bb libraries have a small surface area. There might also be a way to remove more builtin libs and provide them as shared libraries. There's definitely some backwards compatibility and procurement issues that would need to be solved, but it might be a reasonable way to make `bb` smaller and provide access to a large set of libraries at runtime.

**Reply by phronmophobic** [reply]
> Would it be possible to pass function references over the native FFI border for example? I would be surprised if this was possible.

I also doubt that this is possible.

**Reply by borkdude** [reply]
Right. We can just start from a small SCI core and then try to add some namespaces as shared libs and see how that goes.

**Comment by cnuernber** [comment]
Also there is [truffle native interface](https://www.graalvm.org/graalvm-as-a-platform/language-implementation-framework/NFI/)

**Reply by phronmophobic** [reply]
I did not know that existed! That seems better than my approach with libffi via graalvm's native interface. I'll give that a shot and see if it reduces the binary size.

**Reply by phronmophobic** [reply]
Do you know of an example (either java or clj) that uses the truffle native interface? I'm having trouble figuring out how to set things up.

`IllegalArgumentException: A language with id 'nfi' is not installed`

I can't even find an example on the internet 🤷 
https://grep.app/search?q=%20Polyglot.eval%28%27nfi%27
https://grep.app/search?q=eval%28%22nfi%22

It's still marked as experimental, so it might also be worth revisiting this in the future:

> Stability
> 
> The NFI is an internal language designed for language implementors. It is not considered stable and the interface and behavior may change without warning. It is not intended to be used directly by end-users.

**Comment by cnuernber** [comment]
I know you have to include some special arguments to include it - it is a separate module similar to https.  I haven't directly tried to get nfi to work myself.

**Comment by borkdude** [comment]
@phronmophobic Hey, I figured, we could include this behind a feature flag for people who want to have this today.

https://github.com/babashka/babashka/blob/master/doc/build.md#feature-flags

What do you think of that idea?

Also with Java 17 in mind, I wonder if there will be new possibilities using Panama or so.

**Reply by phronmophobic** [reply]
Including as a feature flag makes a lot of sense. There's still some significant caveats to the current version:

- Mac OSX only
- No support for structs
- No support for callbacks

On the bright side, dtype-next should now support clj-easy/graal-build-time, https://github.com/cnuernber/dtype-next/issues/41

**Reply by phronmophobic** [reply]
> Also with Java 17 in mind, I wonder if there will be new possibilities using Panama or so.

I'm pretty excited about the potential for Panama. Hopefully, it means you can get libffi features with native-image for free!

**Reply by borkdude** [reply]
We can mark it as as _very experimental_ so users can use it with these caveats.

**Comment by ieugen** [comment]
From what I could tell, Java 22 will probably include panama final version.
